content/blog/2022/masync-ex-hashmap.md

```diff
@@ -0,0 +1,476 @@
+				goto set_output;
+			case HASHMAP_ENTRY_STATE_UNOCCUPIED:
+			/*
+			 * An unoccupied entry wasn't found despite hashmap not
+			 * being full. Re-run the lookup future.
+			 */
+				return FUTURE_STATE_RUNNING;
+			default:
+				assert(0); /* should not be reachable */
+		}
+	}
+```
+If searching for the index of an existing entry failed, we proceed similarly to the
+failed check case.
+
+If we were looking for the index of an unoccupied entry and nothing was found, we return
+`FUTURE_STATE_RUNNING`. `hashmap_lookup_impl` will be re-run upon the next poll,
+preserving all the changes made to the data and output.
+
+```C
+	hme = &hm->entries[index];
+set_output:
+	output->hme = hme;
+	return FUTURE_STATE_COMPLETE;
+```
+Lastly, when we found an entry, we set the output entry pointer to its address and
+return `FUTURE_STATE_COMPLETE`.
+
+## Initializing a future
+
+Having defined the future input, output and task function, we can create an instance of
+the `lookup` future.
+
+```C
+static struct hashmap_lookup_fut
+hashmap_lookup(struct hashmap *hm, uint64_t key, enum hashmap_entry_state state)
+{
+	struct hashmap_lookup_fut future;
+	/* Set input values */
+	future.data.hm = hm;
+	future.data.key = key;
+	future.data.state = state;
+	/* Set default output value */
+	future.output.hme = NULL;
+	FUTURE_INIT(&future, hashmap_lookup_impl);
+	return future;
+}
+```
+
+In the above code fragment, we create a `struct hashmap_lookup_fut` structure and set its input data and
+default output values. After that, we initialize this future and bind it with a `hashmap_lookup_impl(...)`
+task function using the `FUTURE_INIT(...)` macro. After those operations, `struct hashmap_lookup_fut` is
+ready to be polled.
+
+## Creating a chained future
+
+Multiple futures can be composed into a single chained future. When polled, chained
+future executes its future entries sequentially in the order they were defined.
+We can define a future as the chained future entry using `FUTURE_CHAIN_ENTRY(_future_type, _name)`
+macro. It takes two parameters, `_future_type` is the type of the future, and `_name`
+is the name of the variable stored in the data structure.
+
+```C
+struct hashmap_lookup_lock_entry_data {
+	FUTURE_CHAIN_ENTRY(struct hashmap_lookup_fut, lookup);
+	FUTURE_CHAIN_ENTRY(struct hashmap_entry_set_state_fut, set_state);
+	FUTURE_CHAIN_ENTRY_LAST(struct chain_entries_rerun_fut, entries_rerun);
+	struct future_chain_entry *entriesp[2];
+};
+struct hashmap_lookup_lock_entry_output {
+	struct hashmap_entry *hme;
+};
+FUTURE(hashmap_lookup_lock_entry_fut, struct hashmap_lookup_lock_entry_data,
+		struct hashmap_lookup_lock_entry_output);
+```
+Here, `struct hashmap_lookup_lock_entry_fut` future consists of three other futures,
+`struct hashmap_lookup_fut` that was discussed, `struct hashmap_entry_set_state_fut`,
+and 'struct chain_entries_rerun_fut' that follow the same principles the `lookup` future does.
```

Comment:
I think `struct chain_entries_rerun_fut` was probably supposed to be inside the `  signs.